File: examples/java/AddTagging.java

import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TIOStreamTransport;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import streamcorpus.EntityType;
import streamcorpus.Sentence;
import streamcorpus.StreamItem;
import streamcorpus.Token;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.Arrays;

/**
 * User: jacek
 * Date: 3/8/13
 * Time: 3:38 PM
 */
public final class AddTagging {
    public static void main(String[] args) {
        try {
            final TTransport i_transport = new TIOStreamTransport(new FileInputStream("test-data/john-smith-tagged-by-lingpipe-0-v0_3_0.sc"));
            final TTransport o_transport = new TIOStreamTransport(new FileOutputStream("test-data/john-smith-tagged-by-lingpipe-0-v0_3_0-augmented.sc"));
            try {
                final TBinaryProtocol i_protocol = new TBinaryProtocol(i_transport);
                final TBinaryProtocol o_protocol = new TBinaryProtocol(o_transport);
                i_transport.open();
                int counter = 0;
                while (true) {
                    final StreamItem item = new StreamItem();
                    item.read(i_protocol);
                    System.out.println("counter = " + ++counter);
                    System.out.println("item = " + item);
                    System.out.println("item.body.clean_visible.length " + item.body.clean_visible.length());

                    item.body.sentences.put("my_tagger", Arrays.asList(sentence(tok(0, "The"), tok(1, "cat"), tok(2, "jumped"), tok(3, "high"))));
                    item.write(o_protocol);
                }
            } catch (TTransportException te) {
                if (te.getType() == TTransportException.END_OF_FILE) {
                    System.out.println("*** EOF ***");
                    i_transport.close();
                    o_transport.close();
                } else {
                    throw te;
                }
            }
        } catch (Exception e1) {
            System.err.println(e1);
        }
    }

    // auxiliary functions to make Sentence creation less verbose
    // bringing Java code closer to the spirit of Scala

    static Token tok(int num, String content) {
        final Token tok = new Token(num, content);
        tok.equiv_id = 2;
        tok.entity_type = EntityType.FAC;
        return tok;
    }

    static Sentence sentence(Token... tokens) {
        return new Sentence(Arrays.asList(tokens));
    }
}

File: examples/scala/AddTagging.scala
package test.scala.streamcorpus

import org.apache.thrift.transport.{TTransportException, TIOStreamTransport, TTransport}
import java.io.{FileOutputStream, FileInputStream}
import org.apache.thrift.protocol.TBinaryProtocol
import streamcorpus.{Token, Sentence, StreamItem, EntityType}

/**
 * User: jacek
 * Date: 8/10/13
 * Time: 3:20 PM
 */

object AddTagging {
  def main(args: Array[String]) {
    val i_transport: TTransport = new TIOStreamTransport(new FileInputStream("../../test-data/john-smith-tagged-by-lingpipe-0-v0_3_0.sc"))
    val o_transport: TTransport = new TIOStreamTransport(new FileOutputStream("/tmp/john-smith-tagged-by-lingpipe-0-v0_3_0.sc"))
    val i_binProto: TBinaryProtocol = new TBinaryProtocol(i_transport)
    val o_binProto: TBinaryProtocol = new TBinaryProtocol(o_transport)
    i_transport.open()
    while (true) {
      try {
        val item: StreamItem = StreamItem.decode(i_binProto)
        item.body match {
          case Some(contentItem) => {

	    /* jacek, not sure if this is correct syntax -- can you fix it for me? */
	    val tok1: Token = new Token(0, "The")
	    tok1.token_num = 0

	    val tok2: Token = new  Token(1, "cat")
	    tok1.token_num = 1

	    /* especially this needs to work, see Java example */
	    tok2.entity_type = EntityType.PER
	    
	    val tok3: Token = new  Token(1, "is")
	    tok1.token_num = 2


	    val sentence1: Sentence = new Sentence(Seq(tok1, tok2, tok3))
            contentItem.sentences + ("my_tagger" -> Seq(sentence1))
          }
          case None => println("no content")
        }
        item.write(o_binProto)
      } catch {
        case te: TTransportException => if (te.getType == TTransportException.END_OF_FILE) {
          i_transport.close()
          o_transport.close()
          println("END")
          return
        } else {
          println("type " + te.getType)
        }
      }
    }
  }
}